[baseline-alignment] Move BaselineContext and BaselineGroup out of GridBaselineAlignment.

sammygill opened the pull request:
#### 493db53c8c0b4ad4b3c1180d356af40a679db746
<pre>
[baseline-alignment] Move BaselineContext and BaselineGroup out of GridBaselineAlignment.
<a href="https://bugs.webkit.org/show_bug.cgi?id=256775">https://bugs.webkit.org/show_bug.cgi?id=256775</a>
rdar://problem/109336789

Reviewed by Brent Fulgham and Tim Nguyen.

These two classes represent parts of the baseline alignment spec in
css-align-3 in general and are not specific to grid. These classes
can be useful for baseline alignment in other formatting contexts, such
as flex, so we will attempt to share them between all of the contexts
where needed. We may be able to abstract more things out of GridBaselineAlignment,
but it is not super clear what else at this point so it will be best to
start here, integrate the classes into another formatting context, and
then come back and see what else, if anything, can be moved around.

A couple of new #includes were needed in TextSizeAdjustment.cpp and
RenderGrid.cpp due to new compile errors in the unified sources.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/rendering/BaselineAlignment.cpp: Added.
(WebCore::BaselineGroup::BaselineGroup):
(WebCore::BaselineGroup::update):
(WebCore::BaselineGroup::isOppositeBlockFlow const):
(WebCore::BaselineGroup::isOrthogonalBlockFlow const):
(WebCore::BaselineGroup::isCompatible const):
(WebCore::BaselineContext::BaselineContext):
(WebCore::BaselineContext::sharedGroup const):
(WebCore::BaselineContext::updateSharedGroup):
(WebCore::BaselineContext::findCompatibleSharedGroup):
* Source/WebCore/rendering/BaselineAlignment.h: Copied from Source/WebCore/rendering/GridBaselineAlignment.h.
(WebCore::BaselineGroup::maxAscent const):
(WebCore::BaselineGroup::size const):
* Source/WebCore/rendering/BaselineAlignmentInlines.h: Added.
(WebCore::isBaselinePosition):
(WebCore::isFirstBaselinePosition):
* Source/WebCore/rendering/GridBaselineAlignment.cpp:
(WebCore::BaselineGroup::BaselineGroup): Deleted.
(WebCore::BaselineGroup::update): Deleted.
(WebCore::BaselineGroup::isOppositeBlockFlow const): Deleted.
(WebCore::BaselineGroup::isOrthogonalBlockFlow const): Deleted.
(WebCore::BaselineGroup::isCompatible const): Deleted.
(WebCore::BaselineContext::BaselineContext): Deleted.
(WebCore::BaselineContext::sharedGroup const): Deleted.
(WebCore::BaselineContext::updateSharedGroup): Deleted.
(WebCore::BaselineContext::findCompatibleSharedGroup): Deleted.
* Source/WebCore/rendering/GridBaselineAlignment.h:
(WebCore::BaselineGroup::maxAscent const): Deleted.
(WebCore::BaselineGroup::size const): Deleted.
(): Deleted.
(WebCore::isBaselinePosition): Deleted.
(WebCore::isFirstBaselinePosition): Deleted.
* Source/WebCore/rendering/RenderGrid.cpp:
* Source/WebCore/rendering/style/TextSizeAdjustment.cpp:

Canonical link: <a href="https://commits.webkit.org/264113@main">https://commits.webkit.org/264113@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8d30c7a3c76bf65680635d88faec721d3ddca720

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6651 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6867 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/7049 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8242 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6932 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/6650 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/8082 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6815 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/9825 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6769 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/7409 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/6086 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8361 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/4853 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/6015 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/13892 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/6616 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/6095 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/8828 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6585 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5395 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5981 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1590 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10153 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6354 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mdubet commented on Source/WebCore/rendering/BaselineAlignment.h:
```diff
@@ -0,0 +1,126 @@
+    WritingMode m_blockFlow;
+    ItemPosition m_preference;
+    LayoutUnit m_maxAscent;
+    HashSet<const RenderBox*> m_items;
```
Can we change this to `HashSet<const RenderBox&>` ?

sammygill replied:
Ah yeah the raw pointers are probably not ideal. I wonder if instead we should try to use `WeakPtr` here since the `BaselineAlignment` object doesn't own the renderers?

nt1m replied:
Yeah we should find an alternative to raw pointers here.

brentfulgham replied:
RenderBox& and RenderBox* really only differ in nullability, they are still both basically bare pointers that can lead to badness. It's better to use a `HashSet<WeakPtr<const RenderBox>>` or maybe a `HashSet<CheckedRef<const RenderBox>>`, which will avoid misuse.

mdubet commented on Source/WebCore/rendering/BaselineAlignment.h:
```diff
@@ -0,0 +1,126 @@
+    Vector<BaselineGroup> m_sharedGroups;
+};
+
+enum AllowedBaseLine { FirstLine, LastLine, BothLines };
```
`enum class : uint8_t`

mdubet reviewed (commented):
Considering this is just a refactoring, I'm not sure my comments should actually go in this patch or a follow-up 😬

nt1m commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -35,6 +35,7 @@
 #include "LayoutRepainter.h"
 #include "RenderBoxInlines.h"
 #include "RenderChildIterator.h"
+#include "RenderElementInlines.h"
```
Is this needed?

nt1m commented on Source/WebCore/rendering/style/TextSizeAdjustment.cpp:
```diff
@@ -29,6 +29,7 @@
 #if ENABLE(TEXT_AUTOSIZING)
 
 #include "RenderStyle.h"
+#include "RenderStyleInlines.h"
```
Same question here

sammygill replied:
After adding the new file I ended up getting compile errors in these files. Not super familiar with how our unified sources pick which files to use but I'm guessing it has something to do with that? Or if there is another way to solve that issue I could do that

brentfulgham replied:
It makes sense that the inlines would be needed in these implementation files if we are moving code around. It might just be that the UnifiedBuild files got adjusted (since you added a new file) and it no longer gets the benefit of some include that used to be part of the same compilation unit.

brentfulgham commented on Source/WebCore/rendering/BaselineAlignment.h:
```diff
@@ -0,0 +1,126 @@
+ */
+#pragma once
+
+#include "RenderBox.h"
```
I think you can forward-declare RenderBox instead of including here. You should do so (if it compiles) as it will reduce build churn.

brentfulgham commented on Source/WebCore/rendering/BaselineAlignment.h:
```diff
@@ -0,0 +1,126 @@
+
+enum AllowedBaseLine { FirstLine, LastLine, BothLines };
+
+static inline bool isBaselinePosition(ItemPosition position)
```
Nit: Consider putting these in an `inlines` file that includes the definition for ItemPosition::Baseline, etc.

brentfulgham reviewed (approved):
Looks good! The suggestions about WeakPtr/CheckedRef would be a good follow-up patch to this simple "moving code around" patch.

sammygill committed:
[baseline-alignment] Move BaselineContext and BaselineGroup out of GridBaselineAlignment.
https://bugs.webkit.org/show_bug.cgi?id=256775
rdar://problem/109336789

Reviewed by Brent Fulgham and Tim Nguyen.

These two classes represent parts of the baseline alignment spec in
css-align-3 in general and are not specific to grid. These classes
can be useful for baseline alignment in other formatting contexts, such
as flex, so we will attempt to share them between all of the contexts
where needed. We may be able to abstract more things out of GridBaselineAlignment,
but it is not super clear what else at this point so it will be best to
start here, integrate the classes into another formatting context, and
then come back and see what else, if anything, can be moved around.

A couple of new #includes were needed in TextSizeAdjustment.cpp and
RenderGrid.cpp due to new compile errors in the unified sources.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/rendering/BaselineAlignment.cpp: Added.
(WebCore::BaselineGroup::BaselineGroup):
(WebCore::BaselineGroup::update):
(WebCore::BaselineGroup::isOppositeBlockFlow const):
(WebCore::BaselineGroup::isOrthogonalBlockFlow const):
(WebCore::BaselineGroup::isCompatible const):
(WebCore::BaselineContext::BaselineContext):
(WebCore::BaselineContext::sharedGroup const):
(WebCore::BaselineContext::updateSharedGroup):
(WebCore::BaselineContext::findCompatibleSharedGroup):
* Source/WebCore/rendering/BaselineAlignment.h: Copied from Source/WebCore/rendering/GridBaselineAlignment.h.
(WebCore::BaselineGroup::maxAscent const):
(WebCore::BaselineGroup::size const):
* Source/WebCore/rendering/BaselineAlignmentInlines.h: Added.
(WebCore::isBaselinePosition):
(WebCore::isFirstBaselinePosition):
* Source/WebCore/rendering/GridBaselineAlignment.cpp:
(WebCore::BaselineGroup::BaselineGroup): Deleted.
(WebCore::BaselineGroup::update): Deleted.
(WebCore::BaselineGroup::isOppositeBlockFlow const): Deleted.
(WebCore::BaselineGroup::isOrthogonalBlockFlow const): Deleted.
(WebCore::BaselineGroup::isCompatible const): Deleted.
(WebCore::BaselineContext::BaselineContext): Deleted.
(WebCore::BaselineContext::sharedGroup const): Deleted.
(WebCore::BaselineContext::updateSharedGroup): Deleted.
(WebCore::BaselineContext::findCompatibleSharedGroup): Deleted.
* Source/WebCore/rendering/GridBaselineAlignment.h:
(WebCore::BaselineGroup::maxAscent const): Deleted.
(WebCore::BaselineGroup::size const): Deleted.
(): Deleted.
(WebCore::isBaselinePosition): Deleted.
(WebCore::isFirstBaselinePosition): Deleted.
* Source/WebCore/rendering/RenderGrid.cpp:
* Source/WebCore/rendering/style/TextSizeAdjustment.cpp:

Canonical link: https://commits.webkit.org/264113@main
